core/src/main/scala/akka/kafka/internal/ActorConsumer.scala

```diff
@@ -0,0 +1,256 @@
+    consumer ! KafkaConsumerActor.Subscribe(settings.topics)
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (totalDemand > 0) {
+      if (!buffer.hasNext) {
+        consumer ! KafkaConsumerActor.RequestMessages(tps)
+      }
+      else {
+        val msg = buffer.next()
+        onNext(createMessage(msg))
+        pump()
+      }
+    }
+  }
+
+  override def receive: Receive = working
+
+  def working: Receive = LoggingReceive {
+    case KafkaConsumerActor.Assigned(newTps) =>
+      tps ++= newTps
+      pump()
+    case KafkaConsumerActor.Revoked(newTps) =>
+      tps --= newTps
+      pump()
+    case KafkaConsumerActor.Messages(msgs) =>
+      // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
+      if (buffer.hasNext) {
+        buffer = buffer ++ msgs.asInstanceOf[Iterator[ConsumerRecord[K, V]]]
+      }
+      else {
+        buffer = msgs.asInstanceOf[Iterator[ConsumerRecord[K, V]]]
+      }
+      pump()
+    case Request(_) =>
+      pump()
+    case ActorControl.Stop =>
+      onComplete()
+      context.become(stopped)
+      sender() ! ActorControl.Stopped
+    case Cancel => context.stop(self)
+  }
+
+  def stopped: Receive = LoggingReceive {
+    case KafkaConsumerActor.Assigned(newTps) =>
+    case KafkaConsumerActor.Revoked(newTps) =>
+    case KafkaConsumerActor.Messages(msgs) =>
+    case ActorControl.Stop => sender() ! ActorControl.Stopped
+    case Cancel => context.stop(self)
+  }
+}
+
+object TopicPartitionSourceActor {
+  case class RegisterSubSource(tp: TopicPartition)
+
+  def apply[K, V](settings: ConsumerSettings[K, V])(implicit as: ActorSystem): Source[(TopicPartition, Source[ConsumerRecord[K, V], Control]), Control] = {
+    Source.actorPublisher[(TopicPartition, Source[ConsumerRecord[K, V], Control])](Props(new TopicPartitionSourceActor(settings)))
+      .mapMaterializedValue(ActorControl(_))
+  }
+
+  class SubSourceActor[K, V](mainSource: ActorRef, tp: TopicPartition, kafka: ActorRef)
+      extends ActorPublisher[ConsumerRecord[K, V]] {
+    var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+    override def receive: Receive = normal
+
+    override def preStart(): Unit = {
+      super.preStart()
+      mainSource ! RegisterSubSource(tp)
+    }
+
+    @tailrec
+    private def pump(): Unit = {
+      if (buffer.hasNext && totalDemand > 0) {
+        onNext(buffer.next())
+        pump()
+      }
+    }
+
+    def normal: Receive = LoggingReceive {
+      case KafkaConsumerActor.Messages(msgs) =>
+        val newMessages = msgs.asInstanceOf[Iterator[ConsumerRecord[K, V]]]
+        buffer = buffer ++ newMessages
+        pump()
+      case Request(_) =>
+        pump()
+        if (!buffer.hasNext && totalDemand > 0) {
+          kafka ! KafkaConsumerActor.RequestMessages(Set(tp))
+        }
+      case Cancel => context.stop(self)
+      case ActorControl.Stop => onComplete()
```

Comment:
do you know that this will not stop the actor? there is `onCompleteThenStop` for that